src/libmeasurement_kit/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,345 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+#ifndef SRC_LIBMEASUREMENT_KIT_NEUBOT_DASH_IMPL_HPP
+#define SRC_LIBMEASUREMENT_KIT_NEUBOT_DASH_IMPL_HPP
+
+#include "../common/utils.hpp"
+#include <measurement_kit/http.hpp>
+#include <measurement_kit/mlabns.hpp>
+#include <measurement_kit/neubot.hpp>
+
+#define DASH_MAX_ITERATIONS 15
+#define DASH_SECONDS 2
+#define MAX_NEGOTIATIONS 512
+
+namespace mk {
+namespace neubot {
+namespace dash {
+
+const std::vector<int> &dash_rates(); // Implemented in dash.cpp
+
+static inline size_t select_lower_rate_index(int speed_kbit) {
+    size_t rate_index = 0;
+    while (dash_rates()[rate_index] < speed_kbit &&
+           rate_index < dash_rates().size()) {
+        rate_index++;
+    }
+    if (rate_index > 0) { // Make sure we don't underflow
+        rate_index -= 1;
+    }
+    return rate_index;
+}
+
+template <MK_MOCK_AS(http::request_send, http_request_send),
+          MK_MOCK_AS(http::request_recv_response, http_request_recv_response)>
+void run_loop_(Var<net::Transport> txp, int speed_kbit, std::string auth_token,
+               Var<report::Entry> entry, Settings settings,
+               Var<Reactor> reactor, Var<Logger> logger, Callback<Error> cb,
+               int iteration = 1) {
+
+    if (iteration > DASH_MAX_ITERATIONS) {
+        logger->debug("dash: completed all iterations");
```

Comment:
Note to self: I should probably use the logger's progress mechanism for the dash test.